File: src_python/traitements/definitions/definition_processus.py

from traitements.definitions import definition_transition
from traitements.definitions.definition_tache import DefinitionTache
from traitements.definitions.definition_transition import DefinitionTransition
from utils import fabrique

class DefinitionProcessus:
    def __init__(self, nom: str, description: str):
        self.nom = nom
        self.description = description
        
        self.premiere_tache  = None
        self.taches = []
        self.transitions = []
         
    @property
    def description(self):
        return self._description
    @description.setter
    def description(self, value):   
        self._description = value   

    @property
    def premiere_tache(self):
        return self._premiere_tache
    @premiere_tache.setter
    def premiere_tache(self, value):
        self._premiere_tache = value
    
    def ajouter_premiere_tache(self, tache: DefinitionTache):
        self.premiere_tache = tache
        self.ajouter_tache(tache)
        
    def ajouter_tache(self, tache: DefinitionTache):
        self.taches.append(tache)

    def ajoute_transition(self, tache_source=None, tache_destination=None, *conditions, definition_transition=None):
        """
        As input, pass either:
            definition_transition
        or:
            tache_source, tache_destination, optional conditions
        """
        if definition_transition:
            # Case 1: add existing transition
            self.transitions.append(definition_transition)
            return definition_transition

        if tache_source is None or tache_destination is None:
            raise ValueError("Either pass a definition_transition or both tache_source and tache_destination")

        # Case 2: create a new DefinitionTransition
        condition_transition = conditions[0] if conditions else (lambda t, d: True)
        transition = fabrique.creer_definition_transition(tache_source=tache_source, tache_destination=tache_destination, condition_transition=condition_transition)

        self.transitions.append(transition)
        return transition
    
    def get_trasitions_sortant_de(self, tache: DefinitionTache):
        return [t for t in self.transitions if t.tache_source == tache]
    
    def is_tache_finale(self, tache: DefinitionTache):
        return len(self.get_trasitions_sortant_de(tache)) == 0

        

File: TP2-eneonce-update/src/main/java/ca/uqam/mgl7460/tp1/types/traitements/definitions/DefinitionProcessus.java
package ca.uqam.mgl7460.tp1.types.traitements.definitions;

import java.util.Iterator;

/**
 * Cette interface représnte les MODÈLES de processus, tels que les trois visualiés dans
 * l'énoncé. Un processus a un nom, une description, et consiste en un graphe où les noeuds sont
 * des (définitions de) tâches, et les liens sont des (définitions de) transitions, pouvant être
 * assujetties de conditions. 
 * 
 * On va désigner  * l'une des tâches comme point de départ du processus. Évidemment, c'est une
 * version (trrès) simplifiée de la 'Business Process Modeling Notation' (BPMN) qui fera l'affaire
 * pour les besoins de ce TP. 
 */
public interface DefinitionProcessus {

    /**
     * retourne le nom du processus
     * @return
     */
    public String getNom();

    /**
     * retourne la description du processus (un cours texte qui explique
     * ce qu'il fait)
     * @return
     */
    public String getDescription();

    /**
     * modifie la description du processus
     * @param description
     */
    public void setDescription(String description);
    
    /**
     * retourne la liste des (définitions de) tâches de (la définition du) processus
     * @return
     */
    public Iterator<DefinitionTache> getTaches ();

    /**
     * Ajoute une (définition de) tâche a (la définition du) processus
     * @param definitionTache
     */
    public void ajouteTache(DefinitionTache definitionTache);

    /**
     * ajoute une (définition de) transition à la définition du processus.
     * Rappelez vous qu'un modèle de (définition de) processus est un graphe
     * dont les noeuds sont des définitions de tâches, et dont les liens sont
     * des définition de tâche.
     * @param definitionTransition
     */
    public void ajouteTransition(DefinitionTransition definitionTransition);

    /**
     * Méthode de "commodité" pour rajouter une définition detransition à un modèle/définition
     * de processus. Plutôt que construire une <code>DefinitionTransition</code> et la passer
     * comme argument à <code>void ajouteTransition(DefinitionTransition dt)</code>, on passe
     * les éléments nécessaire pour construire une <code>DefinitionTransition</code>, et on laisse
     * cette méthode créer la code>DefinitionTransition</code> et l'ajouter.
     * 
     * Notez que la condition de transition est optionnelle. Si elle n'est pas fournie, on utilisera comme
     * ConditionTransition une fonction qui retourne toujours vrai.
     * @param tacheSource
     * @param tacheDestination
     * @param conditions
     * @return
     */
    public DefinitionTransition ajouteTransition(DefinitionTache tacheSource, DefinitionTache tacheDestination, ConditionTransition... conditions);

    /**
     * retourne la liste des transitions sortantes d'une (déinition de) tache donnée (paramètre <code>tache</code>)
     * @param tache
     * @return
     */
    public Iterator<DefinitionTransition> getTransitionsSortantesDe(DefinitionTache tache);

    /**
     * retourne la première tâche de la définition du processus
     * @return
     */
    public DefinitionTache getPremiereTache();

    /**
     * modifie la première tâche du processus
     * @param tache
     */
    public void setPremiereTache(DefinitionTache tache);

    /**
     * cette méthode combine les effets <code>ajouteTache(DefinitionTache tache)</code> (qui suppose
     * que <code>tache</code> ne faisait pas encore partie des noeuds) et 
     * <code>setPremiereTache(DefinitionTache tache)</code>, qui marque la tache nouvelle ajoutée comme
     * première tâche 
     * @param tache
     */
    public void ajoutePremiereTache(DefinitionTache tache);

    /**
     * cette méthode détermine, pour une définition de tâche donnée, si la tâche à des "tâches suivantes" ou non,
     * c-a-d, si elle a des transitions sortantes
     * @param tache
     * @return
     */
    public boolean isTacheFinale(DefinitionTache tache);
}
